src/TimeModeling/Types/judiWavefield.jl

```diff
@@ -43,135 +33,78 @@ Constructors
 
-# transpose(jo)
-transpose(A::judiWavefield{vDT}) where vDT =
-	judiWavefield{vDT}(""*A.name*".'",A.n,A.m,A.info,A.dt,A.data)
+####################################################################
+# JOLI conversion
+jo_convert(::Type{T}, jw::judiWavefield{T}, ::Bool) where {T<:Number} = jw
+jo_convert(::Type{T}, jw::judiWavefield{vT}, B::Bool) where {T<:Number, vT} = judiWavefield{T}(jw.nsrc, jv.dt, jo_convert.(T, jw.data, B))
+zero(::Type{T}, v::judiWavefield{vT}; nsrc::Integer=v.nsrc) where {T, vT} = judiWavefield{T}(nsrc, v.dt, T(0) .* v.data[1:nsrc])
+(w::judiWavefield)(x::Vector{<:Array}) = judiWavefield(w.dt, x)
+
+function copy!(jv::judiWavefield, jv2::judiWavefield)
+    v.data .= jv2.data
+    jv.dt = jv2.dt
+    jv
+end
 
-# adjoint(jo)
-adjoint(A::judiWavefield{vDT}) where vDT = conj(transpose(A))
+copyto!(jv::judiWavefield, jv2::judiWavefield) = copy!(jv, jv2)
+make_input(w::judiWavefield) = w.data[1]
+check_compat(ms::Vararg{judiWavefield, N}) where N = all(y -> y.dt == first(ms).dt, ms)
 
+getindex(a::judiWavefield{T}, srcnum::RangeOrVec) where T = judiWavefield{T}(length(srcnum), a.dt[srcnum], a.data[srcnum])
 ####################################################################
 
-function vcat(ai::Vararg{judiWavefield{avDT}, N}) where {avDT, N}
-	N == 1 && (return ai[1])
-	N > 2 && (return vcat(ai[1], vcat(ai[2:end]...)))
-	a, b = ai
-	m = a.m + b.m
-	n = 1
-	nsrc = a.info.nsrc + b.info.nsrc
-	data = Array{Array{avDT, 3}, 1}(undef, nsrc)
-	nt = Array{Integer}(undef, nsrc)
-	for j=1:a.info.nsrc
-		data[j] = a.data[j]
-		nt[j] = a.info.nt[j]
-	end
-	for j=a.info.nsrc+1:nsrc
-		data[j] = b.data[j-a.info.nsrc]
-		nt[j] = b.info.nt[j-a.info.nsrc]
-	end
-	info = Info(a.info.n, nsrc, nt)
-	return judiWavefield{avDT}("judiWavefield",info.n * sum(info.nt), 1, info, a.dt ,data)
-end
-
 function push!(a::judiWavefield{T}, b::judiWavefield{T}) where T
-	append!(a.data, b.data)
-	a.m += b.m
-	a.info.nsrc += b.info.nsrc
+    append!(a.data, b.data)
+    append!(a.dt, b.dt)
+    a.nsrc += b.nsrc
 end
 
 # DFT operator for wavefields, acts along time dimension
-function fft_wavefield(x_in,mode)
-	nsrc = x_in.info.nsrc
-	nt = size(x_in.data[1], 1)
-	if mode==1
-		x = similar(x_in, Complex{Float32})
-		for i=1:nsrc
-			x.data[i] = fft(x_in.data[i], 1)/sqrt(nt)
-		end
-	elseif mode==-1
-		x = similar(x_in, Float32)
-		for i=1:nsrc
-			x.data[i] = real(ifft(x_in.data[i], 1)) * sqrt(nt)
-		end
-	end
-	return x
-end
-
-# Sampling mask to extract wavefields from full vector
-subsample(u::judiWavefield,srcnum) = judiWavefield(u.info,u.data[srcnum];vDT=eltype(u))
-similar(u::judiWavefield) = 0f0 * u
-similar(u::judiWavefield, vDT::DataType) = judiWavefield{vDT}("judiWavefield",u.m, u.n, u.info, u.dt, Array{vDT}.(0f0.*u.data))
-
-# norm
-function norm(a::judiWavefield{avDT}, p::Real=2) where avDT
-    if p == Inf
-        return max([maximum(abs.(a.data[i])) for i=1:a.info.nsrc]...)
+function fft_wavefield(x_in::judiWavefield{T}, mode) where T
```

Comment:
I would overload, i.e. `fft(x_in:judiWavefield)` and `ifft(x_in::judiWavefield)`

Reply:
indeed done